src/optimizer/plugins/AddPathRulesOptimizer.ts

```diff
@@ -0,0 +1,61 @@
+import { Package } from '../../processor';
+import { OptimizerPlugin } from '../OptimizerPlugin';
+import SimplifyArrayIndexingOptimizer from './SimplifyArrayIndexingOptimizer';
+import { ProcessingOptions } from '../../utils';
+import { ExportablePathRule, ExportableSdRule } from '../../exportable';
+import { cloneDeep } from 'lodash';
+import SimplifyRulePathContextsOptimizer from './SimplifyRulePathContextsOptimizer';
+
+export default {
+  name: 'add_path_rules',
+  description: 'Add path rules to support indenting.',
+  runBefore: [SimplifyArrayIndexingOptimizer.name, SimplifyRulePathContextsOptimizer.name],
+  optimize(pkg: Package): void {
+    [
+      ...pkg.profiles,
+      ...pkg.extensions,
+      ...pkg.logicals,
+      ...pkg.resources,
+      ...pkg.instances
+    ].forEach(entity => {
+      let currentRulePath: string;
+      let slicedRulePath: string;
+      let previousSlicedRulePath: string;
+      const seenPathList: string[] = [];
+      const parentList: string[] = [];
+      let entityIndex = -1;
+      const cloneEntity = cloneDeep(entity);
+      cloneEntity.rules.forEach(rule => {
+        currentRulePath = rule.path;
+        entityIndex++;
+        //skip over empty rule paths
+        if (!currentRulePath || currentRulePath === '.') return;
+        //if it's a parent rule path, push it to a seperate list then skip
+        if (!currentRulePath.includes('.')) {
+          if (!parentList.includes(currentRulePath)) parentList.push(currentRulePath);
+          return;
+        }
```

Comment:
I don't understand why we have `parentList` and `seenPathList` -- because they seem to be used in similar ways a few lines down from here.

In fact, there is an issue w/ the current approach because only top-level paths are added to the `parentList` and only `slicedRulePaths` are added to the `seenPathList`.  That means if there is an assignment on a nested rule path (e.g., `a.b`) -- and then assignments on that path's children as well (`a.b.c`, `a.b.d`), then a path rule will be added for `a.b` because it's not in the `parentList` or the `seenPathList` (yet).

I have written two unit tests to show this.  One is on a top-level element, so it passes (thanks to `parentList`), but the other is on a nested path, so it fails.  I'll push those tests up after submitting this review.

I think the best approach would be to have a `seenPathList`, but _every_ rule path should be added to the list (included the path rule paths that are added in the loop). I don't think a separate `parentList` is necessary.

Reply:
4d2d472 removes `parentPath` and adds all paths to `seenPaths`.